File: 85.py
#!/usr/bin/env python

"""This script solves the Project Euler problem "Counting rectangles". The
problem is: Although there exists no rectangular grid that contains exactly two
million rectangles, find the area of the grid with the nearest solution.
"""


def main():
    """Counting rectangles"""

    # Constants
    TARGET = 2000000

    triangle = [0]
    n = 0
    while triangle[n] < TARGET:
        n += 1
        triangle.append(n * (n + 1) / 2)

    closest_area = 0
    closest_diff = TARGET
    for i in range(1, len(triangle)):
        for j in range(i, len(triangle)):
            rectangles = triangle[i] * triangle[j]
            if abs(rectangles - TARGET) < closest_diff:
                closest_diff = abs(rectangles - TARGET)
                closest_area = i * j

    print(closest_area)

if __name__ == '__main__':
    main()

File: 85.js
#!/usr/bin/env node

// Counting rectangles

var TARGET = 2000000;

// Get triangular numbers
var triangle = [0];
var n = 0;
while (triangle[triangle.length - 1] < TARGET) {
    n++;
    triangle.push(n * (n + 1) / 2);
}

var closest_area;
var closest_diff = TARGET;
for (var i = 1; i < triangle.length; i++) {
    for (var j = i; j < triangle.length; j++) {
        var rectangles = triangle[i] * triangle[j];
        if (Math.abs(rectangles - TARGET) < closest_diff) {
            closest_diff = Math.abs(rectangles - TARGET);
            closest_area = i * j;
        }
    }
}

console.log(closest_area);
